### PR TITLE
Add documentation comments to exported TypeScript types

### LINKED ISSUE #758: Make the TypeScript types strict and better
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/got/issues/758)

We have successfully transitioned to TypeScript, but we have not yet enabled strict mode. Although some of the files have strict types, there's still a lot of work to properly type everything.

**If you want to take on this issue, it's expected that you have deep TypeScript understanding and experience. It's not an easy issue.** Comment if you decide to start on this to prevent duplicate work.

I suggest the following order of fixing things (can be multiple PRs):

- [x] Remove `"strict": false,` in tsconfig.json and properly type what's required just to get it to compile (`$ npm test` should pass).
- [x] Improve existing typing and ensure everything has as specific and correct types are possible. (Some progress has been made: https://github.com/sindresorhus/got/pull/760)
- [x] Remove `noImplicitReturns` from tsconfig.json and fix any errors.
- [x] Remove `noUnusedParameters` from tsconfig.json and fix any errors.
- [x] Remove [these lines](https://github.com/sindresorhus/got/blob/0a79ccbe91cc6da1e2e0e41157173e32a5d18a16/package.json#L113-L115) in package.json and fix errors.
- [x] Fix type-related TODO comments.
- [x] Remove `// @ts-ignore` comments and fix the errors.
- [ ] Add doc comments to *public* (meaning exported) functions/types/interfaces. Borrow text from readme. This especially applies to the options and the main Got interface. Follow this style: https://github.com/sindresorhus/typescript-definition-style-guide#documentation
- [ ] Use `readonly` and const assertions wherever possible: https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $180.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($180.00)

### Submitted pull Requests
- [#797 src: Work on strict ts migrations](https://issuehunt.io/r/sindresorhus/got/pull/797)
- [#928 Make TypeScript types conforms to strict mode](https://issuehunt.io/r/sindresorhus/got/pull/928)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/got/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by pmmmwh**
This PR address the last missing bit of #758 by adding doc comments for all public exported types.

Most of the wordings are borrowed from the README.

#### Questions

1. Is there a way to config VS Code/whatever to conform to sindresorhus/typescript-definition-style-guide#documentation? It is a bit painful (and difficult for me to check consistency) when I constantly have to tab/indent comment blocks by hand 🤦 
2. How should I document merged types like `isStream`, `resolveBodyOnly`, `responseType`, `NormalizedOptions`, etc.?

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [ ] I have included some tests.
- [ ] If it's a new feature, I have included documentation updates.

Fixes #758

### CODE REVIEW

**Comment by pmmmwh** [comment]
Hmm ... I'm not sure why tests are failing. All code added in this PR are within comment blocks.

File: source/as-promise/types.ts [hunk 1]
```diff
@@ -72,12 +72,101 @@ export interface Hooks extends RequestHooks {
 
 export interface PaginationOptions<T, R> {
 	pagination?: {
+		/**
+			A function that transform [`Response`](#response) into an array of items.
+			This is where you should do the parsing.
+
+			@default response => JSON.parse(response.body)
+			*/
```
Comment (sindresorhus) [inline source/as-promise/types.ts:80]:
```suggestion
		/**
		A function that transform [`Response`](#response) into an array of items.
		This is where you should do the parsing.

		@default response => JSON.parse(response.body)
		*/
```

Same everywhere else.

**Comment by sindresorhus** [comment]
> Hmm ... I'm not sure why tests are failing.

Just ignore it.

**Comment by sindresorhus** [comment]
Let me know when this is done and ready for review.

**Comment by pmmmwh** [comment]
> Let me know when this is done and ready for review.

I think it is now mostly ready. There are still some parts missing, but they are either:
- Undocumented in the README,
- Or I don't really have an idea of whether they need to be documented, or how they should be documented.

A specific case I have some doubts is for function overloads, like `GotPaginate` or even `Got`/`GotStream`. I'm not sure if TS can correctly pick up the documentation for both overloads without duplicating them ...

**Comment by sindresorhus** [comment]
> I'm not sure if TS can correctly pick up the documentation for both overloads without duplicating them ...

I'm not sure either, but you can test by requiring Got and opening it up in VSCode to see.

**Comment by sindresorhus** [comment]
Can you also update the contribution guidelines to mention that doc changes should also be applied to the doc comments, not just readme.

**Comment by szmarczak** [comment]
> I'm not sure either, but you can test by requiring Got and opening it up in VSCode to see.

I just checked. TS doesn't assume it's the same function with a different name.

**Comment by pmmmwh** [comment]
> > I'm not sure either, but you can test by requiring Got and opening it up in VSCode to see.
> 
> I just checked. TS doesn't assume it's the same function with a different name.

Should I duplicate the docs or leave it alone? I think this issue also exists for re-exports/types that are used in interfaces as values.

(Sorry I haven't been able to dug into this ...)

**Comment by szmarczak** [comment]
> Should I duplicate the docs or leave it alone?

@sindresorhus Maybe let's duplicate them?

**Comment by sindresorhus** [comment]
Yeah, I guess we have to duplicate them.

**Comment by sindresorhus** [comment]
Relevant https://github.com/microsoft/TypeScript/issues/407

**Comment by pmmmwh** [comment]
I finally got time to "finish" this 🤦 

I've duplicated all the documentation as necessary, and have checked that most types should be documented.

I've left out some types that are quite sparse/would rarely be used:

- `OptionsOfTextResponseBody` and its counterparts plus `GotRequestFunction`
- Hook function types (could copy them from the `Hooks` interface) if needed
- Cookie Jars (no applicable documentation as the existing ones only describes the signatures)
- Parts of the `retry.calculateDelay` API (no applicable documentation)
- `NormalizedOptions` and `Defaults`

Please tell me if you want me to fill in any gaps listed above or need to change anything!

**Comment by sindresorhus** [comment]
Thank you for finishing this :)

**Pull request merged**